AR Analysis Template

abeddow91 opened the pull request:
<!-- In this repo you can label a PR with the "PR Deployment" label to deploy the code to a publicly accessible url -->
## What does this change?
Adds and implements a new analysis template to AR. It also adds a new analysis byline component and story and changes the background colour of tags on analysis to white for accessibility.


## Why?
This updates the app analysis design in line with new editorial designs. Design.Analysis previously used the standard layout but the new designs require a reordering of components with some non-standard components used.

## Screenshots

wide
| Before      | After      |
|-------------|------------|
| ![before][] | ![after][] |

[before-wide]: https://user-images.githubusercontent.com/20416599/185168695-cfbf81d8-a589-496c-833e-6bf1ed39da87.png
[after-wide]: https://user-images.githubusercontent.com/20416599/185168768-9465b206-01bf-4847-bed6-fa6edbdb791a.png

mobile
| Before      | After      |
|-------------|------------|
| ![before][] | ![after][] |

[before]: https://user-images.githubusercontent.com/20416599/185168887-b6225c4c-8b94-4eff-a427-14c115e2c97e.png
[after]: https://user-images.githubusercontent.com/20416599/185168931-54e84bd7-0ce5-4ea0-a44d-2f71c5706a5c.png

dark mode -  AR
| Before      | After      |
|-------------|------------|
| ![before-dark][] | ![after-dark][] |

[before-dark]:https://user-images.githubusercontent.com/20416599/185590338-ca900feb-a8df-4a6d-bbe5-e44377725eae.jpeg
[after-dark]: https://user-images.githubusercontent.com/20416599/185590391-f71e5cfb-109a-4699-b2ad-37fab5792f52.png
<!--

You can add extra rows by repeating the last row in the table and then using new unique labels. E.g.

| ![before2][] | ![after2][] |

You can then reference the labels and map them to corresponding links.

[before2]: https://example.com/before2.png
[after2]: https://example.com/after2.png
-->

JamieB-gu commented on apps-rendering/src/components/Byline/AnalysisByline.tsx:
```diff
@@ -0,0 +1,60 @@
+import { darkModeCss } from 'styles';
+import { DefaultByline } from './Byline.defaults';
+
+const blogStyles = css`
```
```suggestion
const styles = css`
```

JamieB-gu commented on apps-rendering/src/components/Byline/AnalysisByline.tsx:
```diff
@@ -0,0 +1,60 @@
+	padding-bottom: ${remSpace[6]};
+`;
+
+const blogAnchorStyles = (format: ArticleFormat) => css`
```
```suggestion
const anchorStyles = (format: ArticleFormat) => css`
```

JamieB-gu commented on apps-rendering/src/components/Byline/AnalysisByline.tsx:
```diff
@@ -0,0 +1,60 @@
+import { DefaultByline } from './Byline.defaults';
+
+const blogStyles = css`
+	${headline.medium({
+		fontWeight: 'light',
+	})}
+
+	color: ${neutral[46]};
+	line-height: 38px;
+	display: flex;
+	flex-direction: column;
+	font-style: italic;
+
+	${until.mobileMedium} {
+		${headline.small({
+			fontWeight: 'light',
+		})}
+	}
```
Can we use the Source API for this?

```suggestion
	${headline.medium({
		fontWeight: 'light',
		fontStyle: 'italic',
		lineHeight: 'tight',
	})}

	color: ${neutral[46]};
	display: flex;
	flex-direction: column;

	${until.mobileMedium} {
		${headline.small({
			fontWeight: 'light',
			fontStyle: 'italic',
			lineHeight: 'tight',
		})}
	}
```

JamieB-gu commented on apps-rendering/src/components/Byline/AnalysisByline.tsx:
```diff
@@ -0,0 +1,60 @@
+		fontWeight: 'light',
+	})}
+
+	color: ${neutral[46]};
```
Can we add this to `editorialPalette` as `text.byline`?

abeddow91 replied:
Definitely, will move this up.

JamieB-gu commented on apps-rendering/src/components/Byline/AnalysisByline.tsx:
```diff
@@ -0,0 +1,60 @@
+	<DefaultByline
+		format={format}
+		bylineHtml={bylineHtml}
+		styles={css(blogStyles)}
+		anchorStyles={css([blogAnchorStyles(format)])}
```
The `css` function isn't required if you just have a single `SerializedStyles`. It's used for composing multiple sets of styles together; for example:

```ts
const compoundStyles = css(stylesOne, stylesTwo, stylesThree);
```

```suggestion
		styles={styles}
		anchorStyles={anchorStyles(format)}
```

JamieB-gu commented on apps-rendering/src/components/Layout/AnalysisLayout.tsx:
```diff
@@ -0,0 +1,83 @@
+				<Tags item={item} />
+			</section>
+		</article>
+		<section css={onwardStyles}>
+			<RelatedContent item={item} />
+		</section>
+		<Footer isCcpa={false} format={item} />
```
Can I suggest pulling the related content and footer outside of the `main` tag? I think the article is the "dominant content" or "central topic" of the page [^1], and the onward content and footer probably fall outside that?

The `ImmersiveLayout` may provide an example of how this could look?

https://github.com/guardian/dotcom-rendering/blob/63733237a4b526b4238c09ec21280a6310a81bc5/apps-rendering/src/components/Layout/ImmersiveLayout.tsx#L83-L108

[^1]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/main

abeddow91 replied:
Makes sense, I'll amend this!

JamieB-gu commented on apps-rendering/src/item.ts:
```diff
@@ -367,150 +374,151 @@ const isPicture = hasTag('type/picture');
-			totalBodyBlocks: content.blocks?.totalBodyBlocks ?? body.length,
-			...itemFields(context, request),
+		(
+			request: RenderingRequest,
+			blockId: Option<string>,
+		): LiveBlog | DeadBlog => {
+			const { content } = request;
+			const body = content.blocks?.body ?? [];
+			const pageSize = content.tags.map((c) => c.id).includes('sport/sport')
+				? 30
+				: 10;
+
+			const parsedBlocks = parseLiveBlocks(context)(body, content.tags);
+			const pagedBlocks = getPagedBlocks(pageSize, parsedBlocks, blockId);
+			return {
+				design:
+					content.fields?.liveBloggingNow === true
+						? ArticleDesign.LiveBlog
+						: ArticleDesign.DeadBlog,
+				blocks: parsedBlocks,
+				pagedBlocks,
+				totalBodyBlocks: content.blocks?.totalBodyBlocks ?? body.length,
+				...itemFields(context, request),
+			};
 		};
-	};
```
As far as I can tell nothing has changed here? Is this diff all indentation? 🤔

abeddow91 replied:
Yep all whitespace changes - I've reverted this as requested.

JamieB-gu commented on apps-rendering/src/item.ts:
```diff
@@ -367,150 +374,151 @@ const isPicture = hasTag('type/picture');
-				...itemFieldsWithBody(context, request),
-			};
-		} else if (isFeature(tags)) {
-			return {
-				design: ArticleDesign.Feature,
-				...itemFieldsWithBody(context, request),
-			};
-		} else if (isLive(tags)) {
-			return fromCapiLiveBlog(context)(request, page);
-		} else if (isRecipe(tags)) {
-			return {
-				design: ArticleDesign.Recipe,
-				...itemFieldsWithBody(context, request),
-			};
-		} else if (isQuiz(tags)) {
-			return {
-				design: ArticleDesign.Quiz,
-				...itemFieldsWithBody(context, request),
-			};
-		} else if (isLabs(tags)) {
 			return {
 				design: ArticleDesign.Standard,
 				...itemFieldsWithBody(context, request),
-				theme: ArticleSpecial.Labs,
 			};
-		} else if (isMatchReport(tags)) {
-			return {
-				design: ArticleDesign.MatchReport,
-				football: parseMatchScores(
-					fromNullable(request.footballContent),
-				),
-				...itemFieldsWithBody(context, request),
-			};
-		}
-
-		return {
-			design: ArticleDesign.Standard,
-			...itemFieldsWithBody(context, request),
 		};
-	};
```
Are there code changes here, or is this all indentation? If this and the above change to `fromCapiLiveBlog` are both just indentation would it be ok if you reverted them? I think it might make this PR easier to review and may cause less confusion for anyone digging through the blame for this file in future.

abeddow91 replied:
Just whitespace changes. I've reverted these and the ones further up so it doesn't muddy the water!

JamieB-gu reviewed (commented):
Nice one integrating with the new sub-component/index/defaults patterns for these new components! I'd be interested to hear your thoughts on this pattern and how easy it is to work with.

A few questions about colour contrast:

1. Do we need to update the colour of the tags against this new background colour?
2. Do we need a dark mode colour for that new "Analysis" line in the standfirst?
3. Does the byline link colour in dark mode meet accessibility guidelines? Firefox is reporting low contrast for that red on black (I haven't been able to test the variants for the other pillars). cc @HarryFischer on this one.

| Tags | Dark Mode |
| - | - |
| ![analysis-tags] | ![analysis-dark] |

[analysis-tags]: https://user-images.githubusercontent.com/53781962/185185262-25d1f04a-8225-4d81-83ef-147914792aca.jpg
[analysis-dark]: https://user-images.githubusercontent.com/53781962/185185271-1878b657-5856-4c1d-a96c-abaed6bf6784.jpg

JamieB-gu reviewed (approved):
Looks good, thanks!

Given that you're merging to a feature branch, perhaps we can merge this and make the contrast/accessibility issue a card on your board that we deal with before that branch gets merged to `main`? That way we get these changes in to unblock other work, but we also make sure we fix any accessibility problems before anything gets to production?

HarryFischer commented:
For me news.500 / #FF5943 (which we currently use in dark mode news articles) passed on the grey background neutral.10 / #1A1A1A. Is that what Firefox is complaining about?

JamieB-gu commented on apps-rendering/src/components/Byline/AnalysisByline.tsx:
```diff
@@ -0,0 +1,60 @@
+`;
+
+const anchorStyles = (format: ArticleFormat) => css`
+	color: ${background.byline(format)};
+
+	${darkModeCss`
+	color: ${background.byline(format)};
```
I've just realised these are using `background.byline` when I think they should probably be using `text.bylineAnchor` and `text.bylineAnchorDark`?

abeddow91 replied:
Totally right, I'll change this

JamieB-gu commented on common-rendering/src/editorialPalette/background.ts:
```diff
@@ -596,6 +596,53 @@ const articleContent = (format: ArticleFormat): string => {
+			case ArticlePillar.Sport:
+				return sport[300];
+			case ArticlePillar.Culture:
+				return culture[300];
+			case ArticlePillar.Opinion:
+				return opinion[300];
+			case ArticlePillar.Lifestyle:
+				return lifestyle[300];
+			case ArticleSpecial.Labs:
+				return labs[300];
+			case ArticleSpecial.SpecialReport:
+				return brandAlt[300];
+			case ArticlePillar.News:
+			default:
+				return news[300];
+		}
+	}
+
+	return neutral[100];
+};
+
+const bylineDark = (format: ArticleFormat): Colour => {
+	if (format.design === ArticleDesign.Analysis) {
+		switch (format.theme) {
+			case ArticlePillar.Sport:
+				return sport[500];
+			case ArticlePillar.Culture:
+				return culture[500];
+			case ArticlePillar.Opinion:
+				return opinion[500];
+			case ArticlePillar.Lifestyle:
+				return lifestyle[500];
+			case ArticlePillar.News:
+			default:
+				return news[500];
+		}
+	}
+
+	return neutral[0];
+};
```
I think you may instead want to add these to `text.bylineAnchor` and `text.bylineAnchorDark`?

abeddow91 replied:
Good spot, thanks! I'll amend this

JamieB-gu reviewed (commented):
> For me news.500 / #FF5943 (which we currently use in dark mode news articles) passed on the grey background neutral.10 / #1A1A1A. Is that what Firefox is complaining about?

Ah, I think this is actually using news.300 / AB0613? If you look at the screenshot above the byline red looks quite different to the bullet points below, which are in the more orangey news.500.

I've added a couple of comments below which I think may cover this.

abeddow91 commented:
@JamieB-gu I've updated this for darkmode using the text palette. I've also updated the background colour of tags in this PR  as it was a small tweak. If you could take another look and see if you're happy that would be 👍
